File: Data Structures/Vector/Java/Vector.java
public class Vector<T> {
    
    private static final int SIZE = 16;
    private int size;
    private int length;
    private T[] array;

    // default constructor
    public Vector() {
        this.array = (T[]) new Object[SIZE];
        this.length = 0;
    }


    // Constructor, with size
    public Vector(int size) {
        // only if size is greater than 0
        if (size > 0) {
            // set the size, init array and length
            this.size = size;
            this.array = (T[]) new Object[this.size];
            this.length = 0;
        } else {
            // error! can't be less than 1
            System.out.println("Error: size cannot be less than 1");
            throw new ArrayIndexOutOfBoundsException();
        }

    }


    // push an item onto the vector
    public void push(T item) {
        // if we are full, resize array
        if (this.length == this.size) {
            resize(2);
        }
        // set item at current position
        this.array[length] = item;
        // increment to next
        this.length++;

    }

    // removes last item from vector and returns it
    public T pop() {
        // get the last item in vector
        T item = array[this.length-1];
        // as long as length is greater than 0
        if (this.length > 0) {
            // decrement length
            this.length--;
            // if len is less than 1/4 size
            if (this.length < this.size / 4)
                // resize the array
                resize(0.5);
            // return item
            return item;
        }
        else
            // return null, b/c no items
            return null;
    }

    // This method returns the object at the index
    public T at(int index) {
        // if out of range, throw exception
        if (index >= this.length)
            throw new IndexOutOfBoundsException();
        else
            // return the object at the index
            return this.array[index];
    }

    // returns the current length of the vector
    public int length() { return this.length; }

    // returns true if the vector is empty
    public boolean isEmpty() { return this.length == 0; }

    // returns the size of the vector
    public int capacity() { return this.size; }

    /**
     * Method returns the index of the element find
     * @param item, the element to find in the vector
     * @return the index of the item, -1 if not found
     */
    public int find(T item) {
        // loop over the array
        for (int i=0; i<this.length; i++) {
            // if item is current
            if (this.array[i] == item)
                // return the index
                return i;
        }
        // fall off array? return -1 for not found
        return -1;
    }

    // insert item at the index supplied
    public void insert(T item, int index) {
        // if the index is out of range, throw error
        if (index >= this.length)
            throw new IndexOutOfBoundsException();
        else {
            // if the current length is the size of array
            // resize the array by a factor of 2
            if (this.length == this.size)
                resize(2);

            // increment the length of the array
            this.length++;

            // everything from index to the right copy
            T[] temp = (T[]) new Object[this.size];

            // init a pointer j, to start at the index supplied
            int j = index;

            // from the index to end, copy array to temp
            for (int i=0; i<temp.length && j<this.array.length; i++)
                temp[i] = this.array[j++];

            // set the new item to the index
            // then increment the index
            this.array[index++] = item;

            // copy back over
            for (int i=0; index<this.array.length; index++)
                this.array[index] = temp[i++];

        }
    }

    // prepends the vector at index 0 with item supplied
    public void prepend(T item) {
        insert(item, 0);
    }

    // deletes item at index supplied
    public void delete(int index) {
        // from the index supplied, shift vector left
        for (int i=index; i<this.length-1; i++)
            this.array[i] = this.array[i+1];
        // decrement the length of the vector
        this.length--;
    }
    
    // removes the object from the vector
    public void remove(T item) {
        
        // while true ??
        // TODO: is this while loop necessary??
        while (true) {
            // find the item
            int i = find(item);
            // if item is not found, break;
            if (i == -1)
                break;
            // delete the item
            delete(i);
        }
    }

    // print out the vector
    public void print() {
        for (int i=0;i<this.length;i++)
            System.out.println(this.array[i]);
    }

    // resize the vector by the factor supplied
    private void resize(double factor) {
        // multiply the size by factor
        this.size *= factor;
        // create a temp array copying over the current array
        T[] temp = this.array.clone();
        // set array to a new array with new size
        this.array = (T[]) new Object[size];
        // copy the temp array into the this.array with it's new size
        System.arraycopy(temp, 0, this.array, 0, temp.length);
    }

}

File: Data Structures/Vector/C#/Vector.cs
using System;

namespace DataStructures.Classes
{
    /// <summary>
    /// This is the Vector class
    /// Uses the array underneath to hold
    /// the data.  Another way is to use
    /// a linked list.
    /// </summary>
    /// <typeparam name="T">Generic Type</typeparam>
    public class Vector<T>
    {
        private const int SIZE = 16;    // Default Vector Size

        private int mSize;              // size of vector
        private T[] vector;             // vector array
        private int mLength;            // number of items in vector

        /// <summary>
        /// Default constructor
        /// </summary>
        public Vector()
        {
            init(SIZE);   
        }

        /// <summary>
        /// Constructor for initial size for vector
        /// </summary>
        /// <param name="size">Size of vector</param>
        public Vector(int size)
        {
            // if array size is less than 0
            // use default size
            init(size <= 0 ? SIZE : size);
        }

        /// <summary>
        /// Constructor to create a vector with an array
        /// </summary>
        /// <param name="array">Array to construct vector from</param>
        public Vector(T[] array)
        {
            // if array size is greater than default, use array size to init
            init(array.Length > SIZE ? array.Length : SIZE);

            // copy over array to vector
            Array.Copy(vector, array, mSize);
        }

        /// <summary>
        /// Pushes new data to the back of the
        /// vector
        /// </summary>
        /// <param name="data">Data to append to end of vector</param>
        /// <exception cref="NotImplementedException"></exception>
        public void PushBack(T data)
        {
            // TODO: Implement resize method
            if (mLength >= mSize -1)
                throw new NotImplementedException($"Need to implement a method to double array size");

            // put new item in vector array
            vector[++mLength] = data;
        }

        /// <summary>
        /// Removes the last item from the vector
        /// </summary>
        /// <returns>Last item in vector</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws if vector is empty</exception>
        public T PopBack()
        {
            // if we are at -1 or less, empty vector throw exception
            if (mLength < 0)
                throw new ArgumentOutOfRangeException($"Vector is empty");
            
            // returns last item from vector, and decrements no. items
            return vector[mLength--];
        }

        /// <summary>
        /// Gets the first item from the vector
        /// </summary>
        /// <returns>First item in vector</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws if vector is empty</exception>
        public T Front()
        {
            // if length is -1 or less, throw error
            if (mLength < 0)
                throw new ArgumentOutOfRangeException($"Veoctor is empty");
            // return the first item in vector
            return vector[0];
        }

        /// <summary>
        /// Gets the last item in the vector
        /// </summary>
        /// <returns>Last item in vector</returns>
        public T Back()
        {
            return vector[mLength];
        }

        
        /// <summary>
        /// Gets the item at specified index
        /// </summary>
        /// <param name="index">location in vector</param>
        /// <returns>Item at index location</returns>
        /// <exception cref="IndexOutOfRangeException">Throws if out of range of vector</exception>
        public T At(int index)
        {
            // if index is greater than len, out of range
            if (index > mLength)
                throw new IndexOutOfRangeException($"Index out of range of vector");

            // return the item at the index
            return vector[index];
        }

        /// <summary>
        /// Inserts data at index supplied
        /// </summary>
        /// <param name="data">value to add to vector</param>
        /// <param name="index">index which to insert into vector</param>
        /// <exception cref="ArgumentOutOfRangeException">Throws if out of range of vector</exception>
        public void Insert(T data, int index)
        {
            // if index is greater than len, out of range
            if (index > mLength) 
                throw new ArgumentOutOfRangeException($"Index out of range of vector");
            
            //TODO: this doesn't check if moving everything over will overwrite end value....

            // loop through vector and move everything
            // from end to the index
            for (int i = mLength - 1; i >= index; i--)
                vector[i + 1] = vector[i];
            // after for loop, put new data into vector at index
            vector[index] = data;
        }

        /// <summary>
        /// Clears the vector
        /// </summary>
        public void Clear()
        {
            // clears the length by resetting
            // the length to -1
            mLength = -1;
            
        }

        /// <summary>
        /// Checks if the vector is empty
        /// </summary>
        /// <returns>True if empty</returns>
        public bool IsEmpty() => mLength == -1;
        
        /// <summary>
        /// Helper method to initialize the vector class
        /// </summary>
        /// <param name="size">size to create array with</param>
        private void init(int size)
        {
            mSize = size;
            vector = new T[mSize];
            mLength = -1;
        }
    }
}